src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs

```diff
@@ -0,0 +1,461 @@
+            // see bug Devdiv:823958
+            if (text.Lines.GetLineFromPosition(startPoint) == text.Lines.GetLineFromPosition(endPoint))
+            {
+                var startToken = root.FindToken(startPoint, findInsideTrivia: true);
+                if (startToken.IsKind(SyntaxKind.OpenBraceToken) &&
+                    (startToken.Parent?.IsInitializerForArrayOrCollectionCreationExpression() == true ||
+                     startToken.Parent is AnonymousObjectCreationExpressionSyntax))
+                {
+                    // Since the braces are next to each other the span to format is everything up to the opening brace start.
+                    endPoint = startToken.SpanStart;
+                }
+            }
+
+            var style = documentOptions.GetOption(FormattingOptions.SmartIndent);
+            if (style == FormattingOptions.IndentStyle.Smart)
+            {
+                // Set the formatting start point to be the beginning of the first word to the left 
+                // of the opening brace location.
+                // skip whitespace
+                while (startPoint >= 0 && char.IsWhiteSpace(text[startPoint]))
+                {
+                    startPoint--;
+                }
+
+                // skip tokens in the first word to the left.
+                startPoint--;
+                while (startPoint >= 0 && !char.IsWhiteSpace(text[startPoint]))
+                {
+                    startPoint--;
+                }
+            }
+
+            var spanToFormat = TextSpan.FromBounds(Math.Max(startPoint, 0), endPoint);
+            var rules = document.GetFormattingRules(braceFormattingIndentationRules, spanToFormat);
+            var result = Formatter.GetFormattingResult(root, SpecializedCollections.SingletonEnumerable(spanToFormat), document.Project.Solution.Workspace, documentOptions, rules, cancellationToken);
+            if (result == null)
+            {
+                return (ImmutableArray<TextChange>.Empty, closingPoint);
+            }
+
+            var newRoot = result.GetFormattedRoot(cancellationToken);
+            var newClosingPoint = newRoot.GetAnnotatedTokens(s_closingBraceSyntaxAnnotation).Single().SpanStart + 1;
+
+            var textChanges = result.GetTextChanges(cancellationToken).ToImmutableArray();
+            return (textChanges, newClosingPoint);
+
+            static async Task<Document> GetDocumentWithAnnotatedClosingBraceAsync(Document document, int closingBraceEndPoint, CancellationToken cancellationToken)
+            {
+                var originalRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var closeBraceToken = originalRoot.FindToken(closingBraceEndPoint - 1);
+                var newCloseBraceToken = closeBraceToken.WithAdditionalAnnotations(s_closingBraceSyntaxAnnotation);
```

Comment:
1. would it be worth putting some asserts hwere that this found an actual code brace as expected?
2. we don't dive into trivia.  is that expected?  is brace completion only outside of trivia?

Reply:
1.  Sure!
2.  the curly brace completion service is only used outside of trivia - curly braces inside interpolated strings use the interpolation service, otherwise there isn't curly brace completion.